p2p/raintree/network.go

```diff
@@ -50,39 +52,64 @@ func NewRainTreeNetwork(addr cryptoPocket.Address, addrBook typesPre2P.AddrBook)
 		log.Println("[ERROR] Error initializing rainTreeNetwork: ", err)
 	}
 
-	return typesPre2P.Network(n)
+	return types2.Network(n)
 }
 
 func (n *rainTreeNetwork) NetworkBroadcast(data []byte) error {
 	return n.networkBroadcastAtLevel(data, n.maxNumLevels, getNonce())
 }
 
-func (n *rainTreeNetwork) networkBroadcastAtLevel(data []byte, level uint32, nonce uint64) error {
-	// This is handled either by the cleanup layer or redundancy layer
-	if level == 0 {
-		return nil
-	}
+func (n *rainTreeNetwork) networkBroadcastAtLevel(data []byte, level int32, nonce uint64) error {
+	var addr1, addr2 cryptoPocket.Address
+	var ok bool
 
-	msg := &typesPre2P.RainTreeMessage{
+	msg := &types2.RainTreeMessage{
 		Level: level,
 		Data:  data,
 		Nonce: nonce,
 	}
+	// This is handled either by the redundancy layer
+	if level == 0 {
+		if n.redundancyLayerEnabled {
+			// redundancy layer is simply one final send to the original +1/3 && -1/3
+			level = n.maxNumLevels
+			// ensure not an echo-chamber
+			msg.Level = -1
+		} else {
+			if err := n.demote(msg); err != nil {
+				log.Println("Error demoting self during RainTree message propagation: ", err)
+			}
+		}
+	}
+
+	// This is handled by the cleanup layer
+	if level == -1 {
+		// cleanup layer is just send left / right
+		// TODO (Team) unhappy path where the left / right nodes are down
+		// (continue to search left and right until you have a hit)
+		addr1, addr2, ok = n.getLeftAndRight()
+		if !ok {
+			return nil
+		}
+	}
+
 	bz, err := proto.Marshal(msg)
 	if err != nil {
 		return err
 	}
 
-	if addr1, ok := n.getFirstTargetAddr(level); ok {
-		if err = n.networkSendInternal(bz, addr1); err != nil {
-			log.Println("Error sending to peer during broadcast: ", err)
-		}
+	if addr1 == nil {
```

Comment:
I think we'll need to revert it to the previous if structure if we move cleanup into a separate function.

I understand that both approaches work the exact same, and have the same performance, but I tink that doing it the other way will provide a clearer path of how the code operates, even if it may be more redundant.

E.g.

// Cleanup
if level == -1:
 callCLeanupLayer()
 return

// Normal Left 
 if addr1 = getFirstTargets() && addr1 != nil:
  SendToAddr1
  
// Normal right
if addr1 = getFirstTargets() && addr1 != nil:
  SendToAddr1